### PR TITLE
108 make a repository of the guessing game

### LINKED ISSUE #104: Create guessing game viewModel
Create view model to track states of the game, current level's etc. This should be update by the guessing game screen composable.

### LINKED ISSUE #108: Make a repository of the guessing game
Create a repository of the guessing game that fetches level data and includes methods to fetch or update data.

### PR DESCRIPTION
**Pull request by Andreasli02**
Did the repository and viewModel for the game in this branch.

closes #108 closes #104

### CODE REVIEW

File: app/src/main/kotlin/com/platform/openemoji/game/GameRepository.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    // Preferences key used to store the current level counter
+    private val levelCounter = intPreferencesKey("currentLevel")
+
+    override suspend fun getLevelCounter(): Flow<Int> {
+        return context?.dataStore?.data?.map { gameData ->
+            // Retrieve the current level counter or default to 0
+            gameData[levelCounter] ?: 0
+        }
+            ?: flowOf(testLevelCounter)
+    }
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/game/GameRepository.kt]: Can't you use .first() to have this return an Int? Then it matches the method name better, and you're not really using the flow capabilities, are you?

File: app/src/main/kotlin/com/platform/openemoji/game/GameViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private val _currentLevel = MutableStateFlow<Level?>(null)
+    val currentLevel = _currentLevel.asStateFlow()
+
+    private val _levelCounter = MutableStateFlow(0)
+    val levelCounter = _levelCounter.asStateFlow()
+
+    init {
+        viewModelScope.launch {
+            _allLevels.value = gameRepository.getAllLevels()
+        }
```
Comment (OleAndreasER) [inline app/src/main/kotlin/com/platform/openemoji/game/GameViewModel.kt:25]: If it turns out you don't need allLevels, make sure to remove this, the state flow in view model and the methods in repository.

File: app/src/test/kotlin/com/platform/openemoji/game/GameRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    private lateinit var repository: GameMockDataRepository
+
+    @Before
+    fun setup() {
+        val testLevels =
+            listOf(
+                Level(
+                    "🤓🤓",
+                    listOf("Nerd", "Clown", "Ole Andreas", "Andreas Li"),
+                    0,
```
Comment (OleAndreasER) [inline app/src/test/kotlin/com/platform/openemoji/game/GameRepositoryTest.kt:18]: 0 -> 3 ?

**Review by OleAndreasER — COMMENTED** [review]
This seems good to me. I wouldn't have bothered testing the repository (or any of the other repositories), since they are temporary, and with testvalues, it's not really testing much that's used by the app. It's a little funny to make the repository and viewmodel before the UI; I think it's best to do them all at once.

File: app/src/main/kotlin/com/platform/openemoji/game/GameRepository.kt (hunk 1 above)
Comment (Andreasli02) [reply app/src/main/kotlin/com/platform/openemoji/game/GameRepository.kt]: 🫡

File: app/src/main/kotlin/com/platform/openemoji/game/GameViewModel.kt (hunk 2 above)
Comment (Andreasli02) [reply app/src/main/kotlin/com/platform/openemoji/game/GameViewModel.kt:25]: 🫡

File: app/src/test/kotlin/com/platform/openemoji/game/GameRepositoryTest.kt (hunk 3 above)
Comment (Andreasli02) [reply app/src/test/kotlin/com/platform/openemoji/game/GameRepositoryTest.kt:18]: 🤓🫵

**Pull request merged**